src/DataExporter.ts

```diff
@@ -51,33 +57,98 @@ export class DataExporter {
+    const result = await this.iModelDb.getMassProperties(requestContext, requestProps);
    
-    this.iModelDb.withPreparedStatement(ecSql, (statement: ECSqlStatement): void => {
-      if (DbResult.BE_SQLITE_ROW === statement.step()) {
-        for (let i = 0; i < statement.getColumnCount(); i++) {
-          header.push(statement.getValue(i).columnInfo.getAccessString());
-        }   
-        const outHeader: string = header.join(";");
-        writeStream.write(`${outHeader}\n`);
-        writeStream.write(`${this.rowToString(statement)}\n`);
-        rowCount++;
+    return result;
+  }
+
+  public async writeQueryResultsToCsv(ecSql: string, fileName: string, options:  Partial<Options> = {}): Promise<void> {
+    const outputFileName: string = path.join(this.outputDir, fileName); 
+    
+    const opts = Object.assign({
+      calculateMassProperties: false,
+      idColumn: 0,
+      idColumnIsJsonArray: false,
+    }, options);
+
+    await this.iModelDb.withPreparedStatement(ecSql, async (statement: ECSqlStatement): Promise<void> => {     
+        if (opts.calculateMassProperties === true) {
+          await this.writeVolume(statement, outputFileName, opts);
+        } else {
+          await this.defaultWrite(statement, outputFileName);
+        }
       }
+    );
+  }
+
+  private async writeVolume(statement: ECSqlStatement, outputFileName: string, options: Options): Promise<void> {
+    const writeStream = fs.createWriteStream(outputFileName);
+    let rowCount = 0;
+    let ids: Id64Array = [];
+    const header: string[] = ["volume","area"];
+
+    let outHeader: string;
+    if (options.idColumnIsJsonArray) {
+      header.push("CodeValue");
+      outHeader = header.join(';');
+    } else {
+      outHeader = this.makeHeader(header, statement)    
+    }
+    writeStream.write(`${outHeader}\n`);
 
-      while (DbResult.BE_SQLITE_ROW === statement.step()) {
+    while (DbResult.BE_SQLITE_ROW === statement.step()) {
+      let result;
+      if(options.idColumnIsJsonArray === true) {
+        const parsedIds: Id64Array = <Id64Array>JSON.parse(statement.getValue(options.idColumn).getString());
+        result = await this.calculateVolume(parsedIds);
+        writeStream.write(`${result.volume};${result.area};${statement.getValue(1).getString()}\n`);      
+      } else {
         const stringifiedRow = this.rowToString(statement);
-        writeStream.write(`${stringifiedRow}\n`);
-        rowCount++;
+        ids.push(statement.getValue(options.idColumn).getId());
+        result = await this.calculateVolume(ids)
+        ids = [];
+        writeStream.write(`${result.volume};${result.area};${stringifiedRow}\n`);
       }
+      ++rowCount;
+    }
+
+    writeStream.on("finish", () => {
+     console.log(`Written ${rowCount} rows to file: ${outputFileName}`);
     });
+
+    writeStream.end();
+  }
+
+  private async defaultWrite(statement: ECSqlStatement, outputFileName: string): Promise<void> {
```

Comment:
I think defaultWrite and writeVolume could be one method and be simple.  The two will be pretty similar if you remove the special handling for the ids as a json array.  As I noted above we can just push all the columns to the csv file same as if it was a normal query.  The only difference is that we're adding a few columns on to the beginning of the line.